File: keras-rl/game_th/PlayerAI.py
import numpy as np

#TODO: fix and remove try
try:
	from .GameHistory import GameHistory
	from .nnet.NNet import NNet
except Exception as e:
	pass
try:
	from GameHistory import GameHistory
	from nnet.NNet import NNet
except Exception as e:
	pass


env_name = "game_th"
weights_filename				= 'tmp/dqn_{}_weights.h5f'.format(env_name)



class PlayerAI():
	def __init__(self,game):
		self._game_history = GameHistory(game.state_shape())
		self.game_state_shape = (1,) + self._game_history.shape_with_history
		self.num_actions = game.get_num_actions()
		nnet = NNet( self.game_state_shape, self.num_actions )
		nnet.load_weights(weights_filename)
		self.nnet = nnet

	def get_input(self,game_state):
		self._game_history.put(game_state)
		game_state_with_history = self._game_history.get()
		observation = np.reshape(game_state_with_history,(1,) + self.game_state_shape)
		actions = self.nnet.get_action(observation)
		action = np.argmax(actions)
		return action
		

File: keras-rl/html/js/game_th/PlayerAI.js
class PlayerAI{
	constructor(state_shape,num_actions,nnet){
		this._game_history = new GameHistory(state_shape)
		this.game_state_shape = [1].concat(this._game_history.shape_with_history)
		this.num_actions = num_actions
		this.nnet = nnet
	}
	get_input(game_state){
		this._game_history.put(game_state)
		let observation = this._game_history.get()

		let actions = this.nnet.get_action(observation,this.game_state_shape)

		let 
		max_action_prop = getMaxOfArray(actions),
		min_action_prop = getMinOfArray(actions)

		let action_props = []
		for (let action_prop of actions){
			action_props.push((action_prop - min_action_prop)/(max_action_prop - min_action_prop))
		}
		return {"action":np_argmax(actions),"action_props":action_props}
	}
}

function np_argmax(array){
	return array.map((x, i) => [x, i]).reduce((r, a) => (a[0] > r[0] ? a : r))[1]
}

function getMaxOfArray(numArray) {
	return Math.max.apply(null, numArray);
}
function getMinOfArray(numArray) {
	return Math.min.apply(null, numArray);
}